Add unknown field support for csharp

anandolee opened the pull request.

jskeet commented on csharp/src/Google.Protobuf.Test/UnknownFieldSetTest.cs:
```diff
@@ -0,0 +1,127 @@
+using NUnit.Framework;
+
+namespace Google.Protobuf
+{
```
Just starting to review this - I'll be sending a full review later. But before I start, a general point: please reformat all the C# code to use spaces rather than tabs for consistency. (If you're using Visual Studio, it's easy to set this as the default, then use Ctrl-E, D to reformat a whole document.)

anandolee replied:
Done

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,340 @@
+		/// Adds a field to the unknown field set. If a field with the same
+		/// number already exists, the two are merged.
+		/// </summary>
+		public UnknownFieldSet MergeField(int number, UnknownField field)
```
This doesn't looks like it needs to be public.

anandolee replied:
done

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,340 @@
+		}
+
+		/// <summary>
+		/// Gets the default UnknownFieldSet instance.
```
We can't sensibly do this while the type is mutable. Rather than talk about this in individual comments, I'll write an overall comment later.

anandolee replied:
Have remove it and use null instead

anandolee commented on csharp/src/Google.Protobuf/CodedInputStream.cs:
```diff
@@ -1271,6 +1271,30 @@ private void SkipImpl(int amountToSkip)
+        /// Sets the preserveUnknownFieldsDefault true. It is added for
+        /// migration. Will be removed once the migration is done.
+        /// </summary>
+        public static void SetPreserveUnknownsDefaultTrue() {
```
The C++ for example, introduced preserve unknown fields in 3.4.0 and the default is drop. 
3.5.0 changed the default to preserve. Release log ask users to use DiscardUnknownFields() if they want to drop it. I will add DiscardUnknownFields() API in C# as well

anandolee commented on csharp/src/Google.Protobuf/CodedInputStream.cs:
```diff
@@ -1271,6 +1271,30 @@ private void SkipImpl(int amountToSkip)
+        /// <summary>
+        /// Gets the preserveUnknownFieldsDefault;
+        /// </summary>
+        public static bool GetPreserveUnknownsDefault() {
```
Done

anandolee commented on csharp/src/Google.Protobuf/Collections/Lists.cs:
```diff
@@ -0,0 +1,84 @@
+			{
+				return false;
+			}
+			IEqualityComparer<T> comparer = EqualityComparer<T>.Default;
```
ack

anandolee commented on csharp/src/Google.Protobuf/Collections/Lists.cs:
```diff
@@ -0,0 +1,84 @@
+			int hash = 31;
+			foreach (T element in list)
+			{
+				hash = hash*29 + element.GetHashCode();
```
Done

anandolee commented on csharp/src/Google.Protobuf/UnknownField.cs:
```diff
@@ -0,0 +1,269 @@
+		/// </summary>
+		public UnknownField()
+		{
+			varintList = new List<ulong>();
```
done

anandolee commented on csharp/src/Google.Protobuf/UnknownField.cs:
```diff
@@ -0,0 +1,269 @@
+		public override int GetHashCode()
+		{
+			int hash = 43;
+			hash = hash*47 + Lists.GetHashCode(varintList);
```
done

anandolee commented on csharp/src/Google.Protobuf/UnknownField.cs:
```diff
@@ -0,0 +1,269 @@
+		/// </summary>
+		public UnknownField Clear()
+		{
+			varintList.Clear();
```
done

anandolee commented:
Yes, preserving unknown fields are happening in all other languages from 3.4.0.  C++ have already changed the default behavior to preserve in 3.5.0  You can see the release log for more info.

For the DefaultInstance issue, I've remove the one in UnknownFieldSet and use null instead. The UnknownField may still need a default one. I don't quit understand your option 1. Do you mean that add a check in each mutate method, just return if it is the DefaultInstance ?

jskeet commented:
I'd suggest leaving the field null by default even if unknown fields are enabled - I'd expect *most* parse operations not to end up finding any unknown fields, so let's avoid the extra allocation.

Instead, if you create a static method in UnknownFieldSet that accepts an existing field set, and creates a new one if necessary, then you can have:

```csharp
unknownFields = UnknownFields.MergeFieldFrom(unknownFields, input)
````

In equality, you can just use

```csharp
Equals(unknownFields, other.unknownFields)
```

... which will call Equals if and only if both values are non-null. Note that this relies on a field set never being empty - null is effectively *the* representation of an empty unknown field set.

I'd also suggest renaming the field to _unknownFields so that it won't conflict with a message field called "unknown_fields".

anandolee commented:
Have removed the DefaultInstance.
Changed unknownFields to _unknownFields in code gen. 
Also changed to use Equals(_unknownFields, other._unknownFields) in code gen.

For unknownFields = UnknownFields.MergeFieldFrom(unknownFields, input) suggestion, I think it is not needed. Unknown fields are most used in code gen which does not need this support

jskeet commented:
> For unknownFields = UnknownFields.MergeFieldFrom(unknownFields, input) suggestion, I think it is not needed. Unknown fields are most used in code gen which does not need this support

I'm not sure what you mean. The point is that that would allow you to keep `_unknownFields` null until it needs to be populated. Anyway, I'll try to review the new code very early next week.

anandolee commented:
Oh, that's what you mean. I've added those methods.
Thanks for all your feedback

jskeet commented:
I think you need to regenerate all the C# files - it won't build at the moment.
If we're using null for a missing field set, do we still need `Clear()` on either `UnknownFieldSet` or `UnknownField`?

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,371 @@
+
+        private UnknownField GetField(int number)
+        {
+            if (lastField != null)
```
Use && here?

```csharp
if (lastField != null && number == lastFieldNumber)
```

anandolee replied:
done

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,371 @@
+            UnknownField existing;
+            if (fields.TryGetValue(number, out existing))
+            {
+                lastField.MergeFrom(existing);
```
Why are we merging an old one with a new one here? Again, this feels like it's left over from a builder mentality. I'd expect to return `existing` here, and only create a new instance if it's new.

I'd also rename the method to GetOrAdd, as otherwise it's odd that it mutates the set in some cases.

anandolee replied:
done

jskeet commented:
Can I check, is this now ready for complete review as far as you're aware?

anandolee commented:
Yes, I think it is ready fro complete review

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        /// Fetches a field by number, returning an empty field if not present.
+        /// Never returns null.
+        /// </summary>
+        public UnknownField this[int number]
```
Creating a new field but not associating it with anything feels misleading to me:

`fieldSet[10].AddVarint(...)` looks like it's always going to modify the set, but it might not.

I'd suggest either removing the indexer entirely, or letting it just delegate to `fields[number]` so it'll throw a KeyNotFoundException if the field isn't present.

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+    ///
+    /// Most users will never need to use this class directly.
+    /// </summary>
+    public sealed partial class UnknownFieldSet : IMessage
```
Does this need to implement IMessage? Is there something that makes that particularly useful, or is it just that it happens to look similar?

Should we have some way of iterating over the fields? I can see that being useful.

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        public override int GetHashCode()
+        {
+            int ret = 1;
+            foreach (KeyValuePair<int, UnknownField> field in fields)
```
Presumably you're using ^ to make the order irrelevant? It might be worth explicitly stating that.

anandolee replied:
Done

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        /// Checks if the unknown field set is initialized. It is never used.
+        /// Returns true to implement the interface.
+        /// </summary>
+        public bool IsInitialized
```
This isn't part of the C# API - let's remove it.

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+
+        private UnknownField GetOrAddField(int number)
+        {
+            if (lastField != null && number == lastFieldNumber)
```
When would lastField be null, but lastFieldNumber be equal to number in an unexpected way?

(In other words, can we just remove the first condition?)

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        /// Adds a field to the set. If a field with the same number already exists, it
+        /// is replaced.
+        /// </summary>
+        public UnknownFieldSet AddField(int number, UnknownField field)
```
Calling this AddField is confusing when .NET dictionaries tend to throw if you try to add an existing key.

AddOrReplaceField perhaps?

Additionally, it looks like this could be internal (even then, only visible for testing.)

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        /// Parse an entire message from <paramref name="input"/> and merge
+        /// its fields into this set.
+        /// </summary>
+        public void MergeFrom(CodedInputStream input)
```
I believe this can be deleted, if we don't need this to implement IMessage. (And I don't think we do - I've locally removed that, and it looks okay.)

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        /// </summary>
+        /// <param name="input">The coded input stream containing the field</param>
+        /// <returns>false if the tag is an "end group" tag, true otherwise</returns>
+        public bool MergeFieldFrom(CodedInputStream input)
```
Make this private? It doesn't look like it's used outside this class.

jskeet commented on csharp/src/Google.Protobuf/UnknownFieldSet.cs:
```diff
@@ -0,0 +1,364 @@
+        /// If a field number exists in both sets, the values in <paramref name="other"/>
+        /// will be appended to the values in this set.
+        /// </summary>
+        public UnknownFieldSet MergeFrom(UnknownFieldSet other)
```
This can be private too (with one test change)

jskeet commented on csharp/src/Google.Protobuf/UnknownField.cs:
```diff
@@ -0,0 +1,269 @@
+        /// </summary>
+        public int GetSerializedSize(int fieldNumber)
+        {
+            int result = 0;
```
Compute the tag size once? No need to do it for each iteration. Then you can change each loop to something like:

```csharp
result += tagSize * varintList.Count;
foreach (ulong value in varintList)
{
    result += CodedOutputStream.ComputeUInt64Size(value);
}
```

jskeet commented on csharp/src/Google.Protobuf/UnknownField.cs:
```diff
@@ -0,0 +1,269 @@
+                foreach (uint value in fixed32List)
+                {
+                    result += CodedOutputStream.ComputeTagSize(fieldNumber);
+                    result += CodedOutputStream.ComputeFixed32Size(value);
```
Surely Fixed32 and Fixed64 always have the same size, don't they? No need to loop at all here.

jskeet commented on csharp/src/Google.Protobuf/UnknownField.cs:
```diff
@@ -0,0 +1,269 @@
+            return result;
+        }
+
+        /// <summary>
```
Currently all of the methods below here can be internal, can't they? (I'm generally in favour of keeping things internal or private until we definitely need them to be public.)

anandolee commented:
It is not required to provide APIs to get Unknown Field. Some languages like c++ have this API, some(like python) do not provide it. We could not provide it at least now

anandolee commented:
Have changed the 3 methods to internal.

Should also clone the unknown fields. I've added the corresponding code and tests. Sorry I missed it.

We don't have a clear timeline for the migration, this is the java version:
https://github.com/google/protobuf/blob/master/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L470

jskeet commented on csharp/src/Google.Protobuf/CodedInputStream.cs:
```diff
@@ -424,7 +424,10 @@ public void SkipLastField()
+        /// <summary>
+        /// Skip a group.
+        /// </summary>
+        public void SkipGroup(uint startGroupTag)
```
Does this need to be public now? We're skipping it within the library, aren't we? Maybe just internal?

anandolee replied:
Done

anandolee commented:
For other languages, we have released a version that have PreserveUnknownsDefault (default false). Thus users can migrate their code. Then release a version with PreserveUnknownsDefault  true and the last step is remove it (None of them have removed PreserveUnknownsDefault).

So we are indeed going to release a version which have PreserveUnknownsDefault.

jskeet commented:
> So we are indeed going to release a version which have PreserveUnknownsDefault.

Then removing the property later will be a breaking change, meaning that the library should be revved to version 4.x to follow semantic versioning. I'm very, very keen on not breaking compatibility like this. Changing the default over time seems reasonable, but I'm nervous over it being part of the public API.

One alternative option would be to make it an internal property that could be controlled via an external mechanism, e.g. an environment variable. That's not _great_ either, but at least it doesn't completely violate compatibility.

anandolee commented:
Just discussed with Jisi. He thinks it is OK to change the behavior directly(always preserve) and add a DiscardUnknownFields(), as C# does not have to support internal usages. Does this make sense?

jskeet commented:
So remove the property, change codegen as if it were always true, and add the method? Makes sense to me.

Unfortunately we can't add the `DiscardUnknownFields` method to the interface itself (again, a breaking change) - that's the next piece of design work, I guess. Maybe make the "just always preserve them" change, get that merged, and then work on discarding?

anandolee commented:
Have removed the DiscardUnknownFields, so the behavior is changed to always preserve.
Sounds good to work on discarding after this merge.

anandolee commented:
ok to test
